### PR TITLE
Update capabilities and log worker metadata after specialization

### PR DESCRIPTION
**Pull request by kshyju**
This is a second set of changes on host side to address https://github.com/Azure/azure-functions-host/issues/8983. This changes follows the proto file changes made in https://github.com/Azure/azure-functions-language-worker-protobuf/pull/90

In placeholder & specialization use case, the worker metadata and capabilities are available only after specialization (env reload response). In this PR, Updating the code path which handles env reload request to refresh the capabilities and log the worker metadata. 

### Pull request checklist

* [X] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [X] Otherwise: I've added my notes to `release_notes.md`
* [X] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [X] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [X] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

File: src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto [hunk 1]
```diff
@@ -238,6 +238,13 @@ message FunctionEnvironmentReloadRequest {
 }
 
 message FunctionEnvironmentReloadResponse {
+  // After specialization, worker sends capabilities & metadata.
```
Comment (kshyju) [inline src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto]: These changes are published in https://github.com/Azure/azure-functions-language-worker-protobuf/releases/tag/v1.7.1-protofile

Comment (liliankasem) [reply src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto]: I'm not a fan of tying proto changes to host changes. Do you mind creating a new PR to bump the proto version?

Comment (liliankasem) [reply src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto]: there's steps for this exact process here https://dev.azure.com/msazure/One/_git/AAPT-Antares-Docs?path=/TeamDocs/FunctionTeamDocs/EngineeringProcesses/ReleaseProcesses/ProtobufReleaseProcess.md&version=GBmaster&line=1&_a=preview&anchor=pulling-protobuf-updates-into-the-worker/host-repo

And here's an example pr! https://github.com/Azure/azure-functions-host/pull/8946

Comment (kshyju) [reply src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto]: Oh Okay. created another PR for the proto update: https://github.com/Azure/azure-functions-host/pull/9023

Will rebase this PR after the other one is merged.

File: release_notes.md [hunk 2]
```diff
@@ -16,4 +16,6 @@
 
 **Release sprint:** Sprint 132
 [ [bugs](https://github.com/Azure/azure-functions-host/issues?q=is%3Aissue+milestone%3A%22Functions+Sprint+132%22+label%3Abug+is%3Aclosed) | [features](https://github.com/Azure/azure-functions-host/issues?q=is%3Aissue+milestone%3A%22Functions+Sprint+132%22+label%3Afeature+is%3Aclosed) ]
-- Update App Service Authentication/Authorization on Linux Consumption from 1.4.17 to 1.5.1 [Release Note](https://github.com/Azure/app-service-announcements/issues/406)
\ No newline at end of file
+- Update App Service Authentication/Authorization on Linux Consumption from 1.4.17 to 1.5.1 [Release Note](https://github.com/Azure/app-service-announcements/issues/406)
+- Upgraded proto files to v1.7.1-protofile release (#9023)
```
Comment (kshyju) [inline release_notes.md:20]: This one is not part of this PR. I forgot to include release_notes update in 9023.

File: src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto (hunk 1 above)
Comment (kshyju) [reply src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto]: Rebased. proto changes are not part of this PR anymore.

**Review by liliankasem — APPROVED** [review]

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 3]
```diff
@@ -395,7 +397,7 @@ internal void WorkerInitResponse(GrpcEvent initEvent)
             UpdateCapabilities(_initMessage.Capabilities);
 
             _isSharedMemoryDataTransferEnabled = IsSharedMemoryDataTransferEnabled();
-            _cancelCapabilityEnabled = !string.IsNullOrEmpty(_workerCapabilities.GetCapabilityState(RpcWorkerConstants.HandlesInvocationCancelMessage));
+            _cancelCapabilityEnabled ??= !string.IsNullOrEmpty(_workerCapabilities.GetCapabilityState(RpcWorkerConstants.HandlesInvocationCancelMessage));
```
Comment (jviau) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:400]: What if the value of this capability is set to a false-y string? `false`, `0`, etc

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 4]
```diff
@@ -413,6 +415,19 @@ internal void WorkerInitResponse(GrpcEvent initEvent)
...
+        {
+            if (workerMetadata == null)
+            {
+                return;
+            }
+
+            workerMetadata.UpdateWorkerMetadata(_workerConfig);
+            var workerMetadataString = workerMetadata.ToString();
+            _metricsLogger.LogEvent(MetricEventNames.WorkerMetadata, functionName: null, workerMetadataString);
+            _workerChannelLogger.LogDebug("Worker metadata: {workerMetadata}", workerMetadataString);
```
Comment (surgupta-msft) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:428]: This may not be applicable for the scenario here but wanted to call out that for one of my previous PRs - [refer](https://github.com/Azure/azure-functions-host/pull/8796#discussion_r987394851), I got to know that adding log lines here has perf impact. So, we can double check if this log is really needed here and not affecting the perf.

**Review by surgupta-msft — APPROVED** [review]

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 3 above)
Comment (surgupta-msft) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:400]: Also, this may not be part of this PR but can we update the name `_cancelCapabilityEnabled` to may be `_invocationCancelCapabilityEnabled` or something on similar lines. Only cancel capability is not communicating the meaning till I saw `RpcWorkerConstants.HandlesInvocationCancelMessage`

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 5]
```diff
@@ -375,13 +380,10 @@ internal void WorkerInitResponse(GrpcEvent initEvent)
...
 
-            if (_initMessage.WorkerMetadata != null)
-            {
-                _initMessage.UpdateWorkerMetadata(_workerConfig);
-                var workerMetadata = _initMessage.WorkerMetadata.ToString();
-                _metricsLogger.LogEvent(MetricEventNames.WorkerMetadata, functionName: null, workerMetadata);
-                _workerChannelLogger.LogDebug("Worker metadata: {workerMetadata}", workerMetadata);
-            }
+            // In placeholder scenario, the capabilities and worker metadata will not be available
+            // until specialization is done (env reload request). So these can be removed from worker init response code path.
```
Comment (surgupta-msft) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:384]: We should check our understanding here with @soninaren as I remember change in flow on placeholder side for the work - selective loading of extensions.

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 3 above)
Comment (kshyju) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:400]: Currently, host code(in various places) checks the existence of the capability instead of the value being truthy. I agree, Ideally it should be a truthy check. but that might involve coordinating with workers to make sure they send the correct value.

Currently dotnet worker sends capability when the value is truthy. Ex: https://github.com/Azure/azure-functions-dotnet-worker/blob/14cc88859a1b7bbf68ee43409a1ad45a6e8362cb/src/DotNetWorker.Grpc/GrpcWorker.cs#L222-L229

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 4 above)
Comment (kshyju) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:428]: This logging was existing code, which I moved to a private method now. So this PR should not make a perf impact as such.

But I get your point. We can definitely discuss removing this logging entry as we have this information getting logged to functions metrics as well. Thoughts @liliankasem ?

Comment (liliankasem) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:428]: I'm not sure if a detector or other teams are using the information from the log somewhere so I can't say it's 100% safe to remove
